### PR TITLE
Implement -Z function-sections=yes|no

### PR DESCRIPTION
**Pull request by nox**
This lets rustc users tweak whether all functions should be put in their own TEXT section, using whatever default value the target defines if the flag is missing.

I'm having fun experimenting with musl libc and trying to implement the start symbol in Rust, that means avoiding code that requires relocations, and AFAIK putting everything in its own section makes the toolchain generate `GOTPCREL` relocations for symbols that could use plain old PC-relative addressing (at least on `x86_64`) if they were all in the same section.

### CODE REVIEW

**Comment by nox** [comment]
I implemented it as a flag because AFAIK that particular setting doesn't require libstd etc to be built the same thing, thus it is a bit unfortunate that it requires building a whole new target currently to change the value. Plus, it would make it easier to check on Windows if we can turn back `function-sections` on by default there if the bugs are gone.

**Comment by jyn514** [comment]
cc @bjorn3, https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/-ffunction-sections

**Review by bjorn3 — APPROVED** [review]

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -904,6 +904,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "force all crates to be `rustc_private` unstable (default: no)"),
     fuel: Option<(String, u64)> = (None, parse_optimization_fuel, [TRACKED],
         "set the optimization fuel quota for a crate"),
+    function_sections: Option<bool> = (None, parse_opt_bool, [TRACKED],
+        "whether each function should go in its own .text section"),
```
Comment (nagisa) [inline compiler/rustc_session/src/options.rs:908]: `.text` is not exactly the most portable name (e.g. machO uses `__text` AFAIR). I suggest dropping the `.` or perhaps the whole `.text`.

Comment (nox) [reply compiler/rustc_session/src/options.rs]: That's a good point, I removed it.

**Comment by nox** [comment]
Any chance I could get a `p=1` here? I can't wait to play around with that and confirm my understanding of that stuff, but as I'm switching jobs and computers and whatnot all I'm doing is building stuff for Linux on my macOS computer and I don't understand how to do that with a custom `rustc`. 😅

**Comment by nagisa** [comment]
Its the next PR in the [homu queue](https://bors.rust-lang.org/queue/rust).

**Pull request merged**